std/json.c3

```diff
@@ -0,0 +1,428 @@
+        }
+
+        node.children:data():push_back(value_node:data());
+
+        if cursor:eat(char("]")) {
+            // We're done.
+            return make_optional(node);
+        }
+
+        if !cursor:eat(char(",")) {
+            node.children:data():deallocate();
+            return make_optional<JSON_Node>();
+        }
+    }
+
+    // Unreachable.
+    return make_optional<JSON_Node>();
+}
+
+function parse_value : (cursor : Cursor&) -> Optional<JSON_Node> = {
+    let optional_next_token : Optional<u8> = cursor:peak();
+    if !optional_next_token:has_value() {
+        return make_optional<JSON_Node>();
+    }
+
+    const next_token : u8 = optional_next_token:data();
+
+    if next_token == char("\"") {
+        return cursor:parse_string();
+    }
+
+    if next_token == char("{") {
+        return cursor:parse_object();
+    }
+
+    if next_token == char("[") {
+        return cursor:parse_array();
+    }
+
+    // Numbers don't have a prefix character, so just try and parse them.
+    let node : Optional<JSON_Node> = cursor:parse_number();
+    if node:has_value() {
+        return node;
+    }
+
+    if cursor:eat(sv("null")) {
+        return make_optional(new_json_node(/* null */ 0));
+    }
+
+    if cursor:eat(sv("true")) {
+        return make_optional(new_json_node(/* true */ 1));
+    }
+
+    if cursor:eat(sv("false")) {
+        return make_optional(new_json_node(/* false */ 2));
+    }
+
+    return make_optional<JSON_Node>();
+}
+
+function parse_element : (cursor : Cursor&) -> Optional<JSON_Node> = {
+    cursor:eat_whitespace();
+
+    const value_node : Optional<JSON_Node> = cursor:parse_value();
+
+    cursor:eat_whitespace();
+
+    return value_node;
+}
+
+function parse_string : (cursor : Cursor&) -> Optional<JSON_Node> = {
+    if !cursor:eat(char("\"")) {
+        return make_optional<JSON_Node>();
+    }
+
+    const start_pos : isize = cursor.position;
+
+    while !cursor:is_eof() {
+        if cursor:eat(char("\"")) {
+            const end_pos : isize = cursor.position - 1;
+
+            let node : JSON_Node = new_json_node(/* string */ 3);
+            node.value = make_optional(
+                cursor.data:slice_between(start_pos, end_pos)
+            );
+
+            return make_optional(node);
+        }
+
+        // TODO escape characters.
+
+        cursor:next();
+    }
+
+    return make_optional<JSON_Node>();
+}
+
+function is_digit_helper : (token : Optional<u8>) -> bool = {
+    return token:has_value() and ascii_is_digit(token:data());
+}
+
+function parse_number : (cursor : Cursor&) -> Optional<JSON_Node> = {
+    const start_pos : isize = cursor.position;
+    let is_valid : bool = false;
+
+    // This is optional, so discard the return value.
+    cursor:eat(char("-"));
+
+    if !cursor:eat(char("0")) {
+        while is_digit_helper(cursor:peak()) {
+            is_valid = true;
+            cursor:next();
+        }
+    }
+    else {
+        is_valid = true;
+    }
```

Comment:
Should this be `is_valid=false`? I'm not what you're checking for here

Reply:
The code's correct, sorry it's just a bit confusing. If you check the grammar, https://www.json.org/img/number.png, you'll see that if the first digit is zero, then it can only be followed by the decimal point or the exponent. If you have a non-zero digit, then it can be followed by as many other digits as you want.

The `is_valid` flag is there just to ensure that we've parsed at least one digit. We can't use the cursor cause a single `-` shouldn't count as a valid number.